[v14] Update rust to latest stable (1.77.0)

ibeckermayer opened the pull request:
Roughly a backport for https://github.com/gravitational/teleport/pull/39895 and https://github.com/gravitational/teleport/pull/39981, however the rust codebase have diverged enough between this branch and `master` that I am considering this its own thing.

Confirm that build succeeds:
- [ ] https://github.com/gravitational/teleport.e/actions/runs/8474420040

changelog: updates rust to `1.77.0`

Commit:
Updates `RUST_VERSION` to the latest stable (1.77.0)

Commit:
Fixes linting errors

Commit:
Merge branch 'branch/v14' into isaiah/update-rust-branch/v14

camscale commented:
@ibeckermayer Do you mind fixing the lint errors? I usually wait to review until CI passes, so this has been on my list for a couple of weeks now...

Commit:
Comment out #[allow(clippy::arc_with_non_send_sync)]

This is a temporary hack to get CI to pass. It is necessary for the
RUST_VERSION which we are updating to (1.77.0), however CI is still
using an image with the previous version 1.71.1. Meanwhile `arc_with_non_send_sync`
was added in 1.72.0 (https://rust-lang.github.io/rust-clippy/master/index.html#/arc_with_non_send_sync).

The PR with this commit will therefore require a follow up PR to uncomment
the line once the CI image is updated.

ibeckermayer commented:
@camscale see https://github.com/gravitational/teleport/pull/39990#issuecomment-2057697367
